onnx/defs/schema.h

```diff
@@ -438,8 +438,8 @@ class OpSchema final {
           required(false),
           default_value(std::move(default_value_)) {}
 
-    const std::string name;
-    const std::string description;
+    std::string name;
+    std::string description;
```

Comment:
But why?

Reply:
ChatGPT says:

-------

Yes, using placement new to reinitialize an object at the same memory location can be a good option in some cases. However, you need to be careful when using placement new with const fields.

If your OpSchema class has const fields, you cannot reassign their values directly. Instead, you need to use a constructor that initializes those fields during object construction. One way to achieve this is to use an initialization list in the constructor.

Here's an example that demonstrates how to use placement new to reinitialize an object of type OpSchema at the same memory location:

```c++
void resetOpSchema(OpSchema& self, const OpSchema& schema) {
    // Call destructor to destroy the existing object
    self.~OpSchema();

    // Use placement new to construct a new object at the same memory location
    new (&self) OpSchema(schema);
}
```

In this example, the resetOpSchema function takes a reference to an existing OpSchema object (self) and a reference to a const OpSchema object (schema). It first calls the destructor of OpSchema to destroy the existing object at the memory location pointed to by self.

Next, it uses placement new to construct a new OpSchema object at the same memory location. The new operator calls the constructor of OpSchema and passes schema as its argument. The constructor initializes all non-const fields of the OpSchema object, including any const fields that are initialized in the constructor's initialization list.

Note that this technique assumes that the existing object at the memory location pointed to by self was previously constructed using placement new. If the object was constructed using other means, you need to make sure that its memory allocation and alignment match the requirements of the OpSchema class.

Reply:
I asked: Should I just remove the const fields instead?

Reply:
It replied

Whether or not to remove const fields from the OpSchema class depends on the requirements of your application. Removing const fields can make it easier to reinitialize an object using placement new, but it may also affect the semantics of the class.

If the const fields are there to enforce invariants or prevent unintended modifications to the object, removing them may compromise the correctness of your code. On the other hand, if the const fields are not critical to the behavior of the class and removing them does not affect the correctness of your code, then it may be a reasonable option.

If you need to reinitialize an object of a class that has const fields, you can consider using a copy constructor or a move constructor to create a new object and then swap it with the existing object. This approach avoids the need to reinitialize the existing object and can be more efficient than copying the entire object using placement new.

Here's an example that demonstrates how to use a copy constructor to create a new OpSchema object and swap it with the existing object:

```c++
void resetOpSchema(OpSchema& self, const OpSchema& schema) {
    // Create a temporary object using the copy constructor
    OpSchema temp(schema);

    // Swap the temporary object with the existing object
    std::swap(self, temp);
}
```

In this example, the resetOpSchema function creates a temporary OpSchema object temp using the copy constructor, which initializes all fields of the object. It then uses std::swap to exchange the contents of self and temp. The effect is to reinitialize self with the contents of temp, while leaving temp uninitialized. This approach can be more efficient than copying the entire object using placement new, especially if the class has many fields or if the const fields are expensive to copy.

Reply:
Pybind11 should try to map std::map to a dictionary. Maybe you can manually create a py::dict to export the attributes. That should bypass some of the pybind11 templates.

Reply:
Thanks for the suggestion. I think I found a way